File: frontend/src/api/moneyLent.ts
import API from './api';

export interface MoneyLent {
  id: number;
  borrower_name: string;
  amount_lent: number;
  amount_returned: number;
  currency: string;
  currency_display: string;
  date_lent: string;
  expected_return_date?: string;
  status: string;
  status_display: string;
  borrower_contact?: string;
  purpose?: string;
  notes?: string;
  amount_lent_in_inr: number;
  amount_returned_in_inr: number;
  outstanding_amount: number;
  outstanding_amount_in_inr: number;
  return_percentage: number;
  created_at: string;
  updated_at: string;
}

export const fetchMoneyLent = async (): Promise<MoneyLent[]> =>
  (await API.get('/money-lent/')).data;

export const createMoneyLent = async (data: Omit<MoneyLent, 'id' | 'created_at' | 'updated_at' | 'amount_lent_in_inr' | 'amount_returned_in_inr' | 'outstanding_amount' | 'outstanding_amount_in_inr' | 'return_percentage' | 'currency_display' | 'status_display'>): Promise<MoneyLent> =>
  (await API.post('/money-lent/', data)).data;

export const updateMoneyLent = async (id: number, data: Partial<MoneyLent>): Promise<MoneyLent> =>
  (await API.put(`/money-lent/${id}/`, data)).data;

export const deleteMoneyLent = async (id: number): Promise<void> =>
  (await API.delete(`/money-lent/${id}/`)).data;

export const getMoneyLentStatuses = async (): Promise<Array<[string, string]>> =>
  (await API.get('/money-lent/statuses/')).data;

export const getMoneyLentSummary = async () =>
  (await API.get('/money-lent/summary/')).data;

export const getOverdueMoneyLent = async (): Promise<MoneyLent[]> =>
  (await API.get('/money-lent/overdue/')).data;

export const recordPayment = async (id: number, paymentAmount: number) =>
  (await API.post(`/money-lent/${id}/record_payment/`, { payment_amount: paymentAmount })).data;

File: mobile_app/lib/models/money_lent.dart
class MoneyLent {
  final int id;
  final String borrowerName;
  final double amountLent;
  final double amountReturned;
  final String currency;
  final String currencyDisplay;
  final DateTime dateLent;
  final DateTime? expectedReturnDate;
  final String status;
  final String statusDisplay;
  final String? borrowerContact;
  final String? purpose;
  final String? notes;
  final double amountLentInInr;
  final double amountReturnedInInr;
  final double outstandingAmount;
  final double outstandingAmountInInr;
  final double returnPercentage;
  final DateTime createdAt;
  final DateTime updatedAt;

  MoneyLent({
    required this.id,
    required this.borrowerName,
    required this.amountLent,
    required this.amountReturned,
    required this.currency,
    required this.currencyDisplay,
    required this.dateLent,
    this.expectedReturnDate,
    required this.status,
    required this.statusDisplay,
    this.borrowerContact,
    this.purpose,
    this.notes,
    required this.amountLentInInr,
    required this.amountReturnedInInr,
    required this.outstandingAmount,
    required this.outstandingAmountInInr,
    required this.returnPercentage,
    required this.createdAt,
    required this.updatedAt,
  });

  factory MoneyLent.fromJson(Map<String, dynamic> json) {
    return MoneyLent(
      id: json['id'],
      borrowerName: json['borrower_name'],
      amountLent: json['amount_lent'].toDouble(),
      amountReturned: json['amount_returned'].toDouble(),
      currency: json['currency'],
      currencyDisplay: json['currency_display'],
      dateLent: DateTime.parse(json['date_lent']),
      expectedReturnDate: json['expected_return_date'] != null 
          ? DateTime.parse(json['expected_return_date']) : null,
      status: json['status'],
      statusDisplay: json['status_display'],
      borrowerContact: json['borrower_contact'],
      purpose: json['purpose'],
      notes: json['notes'],
      amountLentInInr: json['amount_lent_in_inr'].toDouble(),
      amountReturnedInInr: json['amount_returned_in_inr'].toDouble(),
      outstandingAmount: json['outstanding_amount'].toDouble(),
      outstandingAmountInInr: json['outstanding_amount_in_inr'].toDouble(),
      returnPercentage: json['return_percentage'].toDouble(),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'borrower_name': borrowerName,
      'amount_lent': amountLent,
      'amount_returned': amountReturned,
      'currency': currency,
      'date_lent': dateLent.toIso8601String().split('T')[0],
      'expected_return_date': expectedReturnDate?.toIso8601String().split('T')[0],
      'status': status,
      'borrower_contact': borrowerContact,
      'purpose': purpose,
      'notes': notes,
    };
  }

  bool get isOverdue {
    if (expectedReturnDate == null || status == 'fully_returned') return false;
    return expectedReturnDate!.isBefore(DateTime.now());
  }
}